src/control/server/storage_scm.go

```diff
@@ -57,17 +114,194 @@ var (
+// regions/sets hosting pmem kernel device namespaces.
+//
+// Presents of nonvolatile memory modules is assumed in this method and state
+// is established based on presence and free capacity of regions.
+//
+// Actions based on state:
+// * modules exist and no regions -> create all regions (needs reboot)
+// * regions exist and free capacity -> create all namespaces
+// * regions exist but no free capacity -> no-op
+//
+// Command output from external tools will be returned.
+func (s *scmStorage) Prep() (needsReboot bool, pmemDevs []pmemDev, err error) {
+	if err := s.getState(); err != nil {
+		return false, nil, errors.WithMessage(err, "establish scm state")
+	}
+
+	log.Debugf("scm in state %s\n", s.state)
+
+	switch s.state {
+	case scmStateNoRegions:
+		needsReboot, err = s.createRegions()
+	case scmStateFreeCapacity:
+		pmemDevs, err = s.createNamespaces()
+	case scmStateNoCapacity:
+		pmemDevs, err = s.getNamespaces()
+	default:
+		err = errors.New("unknown scm state")
+	}
+
+	return
+}
+
+// reset executes commands to remove namespaces and regions on SCM models.
+func (s *scmStorage) PrepReset() error {
+	return nil // TODO
+}
+
+// getState establishes state of SCM regions and namespaces on local server.
+func (s *scmStorage) getState() error {
+	s.state = scmStateUnknown
+
+	// TODO: discovery should provide SCM region details
+	out, err := s.runCmd(cmdScmShowRegions)
+	if err != nil {
+		return err
+	}
+
+	if out == outScmNoRegions {
+		s.state = scmStateNoRegions
+		return nil
+	}
+
+	ok, err := hasFreeCapacity(out)
+	if err != nil {
+		return err
+	}
+	if ok {
+		s.state = scmStateFreeCapacity
+		return nil
+	}
+	s.state = scmStateNoCapacity
+
+	return nil
+}
+
+// hasFreeCapacity takes output from ipmctl and checks for free capacity.
+//
+// external tool commands return:
+// $ ipmctl show -d PersistentMemoryType,FreeCapacity -region
+//
+// ---ISetID=0x2aba7f4828ef2ccc---
+//    PersistentMemoryType=AppDirect
+//    FreeCapacity=3012.0 GiB
+// ---ISetID=0x81187f4881f02ccc---
+//    PersistentMemoryType=AppDirect
+//    FreeCapacity=3012.0 GiB
+//
+// FIXME: implementation to be replaced by using libipmctl directly through bindings
+func hasFreeCapacity(text string) (hasCapacity bool, err error) {
+	lines := strings.Split(text, "\n")
+	if len(lines) < 4 {
+		return false, errors.Errorf("expecting at least 4 lines, got %d",
+			len(lines))
+	}
+
+	for _, line := range lines {
+		entry := strings.TrimSpace(line)
+
+		kv := strings.Split(entry, "=")
+		if len(kv) != 2 {
+			continue
+		}
+
+		if kv[0] == "PersistentMemoryType" && kv[1] == "AppDirect" {
+			hasCapacity = true
+			continue
+		}
+
+		if kv[0] != "FreeCapacity" {
+			continue
+		}
+
+		if hasCapacity && kv[1] != "0.0 GiB" {
+			return
+		}
+
+		hasCapacity = false
+	}
+
+	return
+}
+
+// createRegions sets DCPM modules into regions in interleaved AppDirect mode.
+//
+// External tool command output will indicate whether a subsequent reboot is needed.
```

Comment:
Maybe something has changed but last I knew, a reboot was always required after a "create -goal" call, assuming the goal was successfully written. May not be necessary to parse a string to try to figure this out.

Reply:
addressed in https://jira.hpdd.intel.com/browse/DAOS-2997 / https://github.com/daos-stack/daos/pull/854/commits/d0f2c13bd5e456c0075a25af2289cab243a8bf77